Libraries/Opc.Ua.Server/Subscription/MonitoredItem.cs

```diff
@@ -1079,6 +1079,89 @@ public virtual void QueueEvent(EventFieldList fields)
             }
         }
 
+        /// <summary>
+        /// Determines whether an event can be sent with SupportsFilteredRetain in consideration.
+        /// </summary>
+        /// <returns></returns>
+        protected bool CanSendFilteredAlarm(FilterContext context, EventFilter filter, IFilterTarget instance)
+        {
+            bool passedFilter = filter.WhereClause.Evaluate(context, instance);
+
+            ConditionState alarmCondition = null;
+            NodeId conditionId = null;
+            InstanceStateSnapshot instanceStateSnapshot = instance as InstanceStateSnapshot;
+            if (instanceStateSnapshot != null)
+            {
+                alarmCondition = instanceStateSnapshot.Handle as ConditionState;
+
+                if (alarmCondition != null &&
+                    alarmCondition.SupportsFilteredRetain != null &&
+                    alarmCondition.SupportsFilteredRetain.Value &&
+                    filter.SelectClauses != null)
+                {
+                    int lastIndex = filter.SelectClauses.Count - 1;
+                    object value = instance.GetAttributeValue(
+                        context,
+                        filter.SelectClauses[lastIndex].TypeDefinitionId,
+                        filter.SelectClauses[lastIndex].BrowsePath,
+                        filter.SelectClauses[lastIndex].AttributeId,
+                        filter.SelectClauses[lastIndex].ParsedIndexRange);
+
+                    if (value != null)
+                    {
+                        conditionId = value as NodeId;
+                    }
+                }
+            }
+
+            bool canSend = passedFilter;
+
+            // ConditionId is valid only if FilteredRetain is set for the alarm condition
```

Comment:
Why is the ConditionId only valid if FilteredRetain is set ?

Reply:
The logic is specifically for SupportsFilterRetain, so if it is not set, the remainder of the function is not required, only the filter check at the top of the function.

This is still a work in progress, I'm going to move it to draft for now.